python/ray/autoscaler/aws/config.py

```diff
@@ -332,15 +340,75 @@ def _get_vpc_id_or_die(config, subnet_id):
 
 
 def _get_security_group(config, vpc_id, group_name):
+    security_group = _get_security_groups(config, [vpc_id], [group_name])
+    return None if not security_group else security_group[0]
+
+
+def _get_security_groups(config, vpc_ids, group_names):
     ec2 = _resource("ec2", config)
     existing_groups = list(
         ec2.security_groups.filter(Filters=[{
             "Name": "vpc-id",
-            "Values": [vpc_id]
+            "Values": vpc_ids
         }]))
+    matching_groups = []
     for sg in existing_groups:
-        if sg.group_name == group_name:
-            return sg
+        if sg.group_name in group_names:
+            matching_groups.append(sg)
+    return matching_groups
+
+
+def _create_security_group(config, vpc_id, group_name):
+    client = _client("ec2", config)
+    client.create_security_group(
+        Description="Auto-created security group for Ray workers",
+        GroupName=group_name,
+        VpcId=vpc_id)
+    security_group = _get_security_group(config, vpc_id, group_name)
+    logger.info("_create_security_group: Created new security group {} ({})"
+                .format(security_group.group_name, security_group.id))
+    assert security_group, "Failed to create security group"
+    return security_group
+
+
+def _upsert_security_group_rules(conf, security_groups):
+    head_sg = security_groups["head"]
+    worker_sg = security_groups["workers"]
+
+    sgids = {sg.id for sg in security_groups.values()}
+    _update_inbound_rules(head_sg, sgids, conf)
+    _update_inbound_rules(worker_sg, sgids, conf)
+
+
+def _update_inbound_rules(target_security_group, sgids, config):
+    if not target_security_group.ip_permissions:
+        IpPermissions = [{
+            "FromPort": -1,
+            "ToPort": -1,
+            "IpProtocol": "-1",
+            "UserIdGroupPairs": [
+                {
+                    "GroupId": security_group_id
+                } for security_group_id in sorted(sgids)
+                # sort security group IDs for deterministic IpPermission models
+                # (mainly supports more precise stub-based boto3 unit testing)
+            ]
+        }]
+
+        IpPermissions.extend([{
+            "FromPort": channel[FROM_PORT],
+            "ToPort": channel[TO_PORT],
+            "IpProtocol": channel[IP_PROTOCOL],
+            "IpRanges": [{
+                "CidrIp": "0.0.0.0/0"
+            }]
+        } for channel in DEFAULT_INBOUND_CHANNELS])
+
+        additional_IpPermissions = config["provider"].get(
+            "security_group", {}).get("IpPermissions", [])
+        IpPermissions.extend(additional_IpPermissions)
```

Comment:
For this, how about we pull this out into a pure function that constructs this data structure? a la `default_security_group_flows(trusted_sgs, extended_rules=None)`.

Then we can also namespace the constants out of the module.

Reply:
Also does it maybe make sense for them to be explicitly marked as private?

Reply:
The pure function to construct IpPermissions sounds good. The inbound rule whitelist pull request that will follow this PR does something very similar, so this represents an early step in the right direction.

I think it makes sense to mark any remaining newly introduced constants as private (using the leading underscore naming convention). The only other place that they're used is within unit tests, and other external users should probably be discouraged from taking any non-test dependencies on them.